zarr/storage.py

```diff
@@ -1877,6 +1878,204 @@ def __delitem__(self, key):
             self._invalidate_value(key)
 
 
+class SQLiteStore(MutableMapping):
+    """Storage class using SQLite.
+
+    Parameters
+    ----------
+    path : string
+        Location of database file.
+    **kwargs
+        Keyword arguments passed through to the `sqlite3.connect` function.
+
+    Examples
+    --------
+    Store a single array::
+
+        >>> import zarr
+        >>> store = zarr.SQLiteStore('data/array.sqldb')
+        >>> z = zarr.zeros((10, 10), chunks=(5, 5), store=store, overwrite=True)
+        >>> z[...] = 42
+        >>> store.close()  # don't forget to call this when you're done
+
+    Store a group::
+
+        >>> store = zarr.SQLiteStore('data/group.sqldb')
+        >>> root = zarr.group(store=store, overwrite=True)
+        >>> foo = root.create_group('foo')
+        >>> bar = foo.zeros('bar', shape=(10, 10), chunks=(5, 5))
+        >>> bar[...] = 42
+        >>> store.close()  # don't forget to call this when you're done
+    """
+
+    def __init__(self, path, **kwargs):
+        import sqlite3
+
+        kwargs.setdefault('timeout', 5.0)
+        kwargs.setdefault('cached_statements', 100)
+
+        # normalize path
+        if path != ':memory:':
+            path = os.path.abspath(path)
+
+        # store properties
+        self.path = path
+        self.kwargs = kwargs
+
+        # allow threading if SQLite connections are thread-safe
+        #
+        # ref: https://www.sqlite.org/releaselog/3_3_1.html
+        # ref: https://bugs.python.org/issue27190
+        check_same_thread = True
+        if sqlite3.sqlite_version_info >= (3, 3, 1):
+            check_same_thread = False
+
+        # open database
+        self.db = sqlite3.connect(
+            self.path,
+            detect_types=0,
+            isolation_level=None,
+            check_same_thread=check_same_thread,
+            **self.kwargs
+        )
+
+        # handle keys as `str`s
+        self.db.text_factory = str
+
+        # get a cursor to read/write to the database
+        self.cursor = self.db.cursor()
+
+        # initialize database with our table if missing
+        self.cursor.execute(
+            'CREATE TABLE IF NOT EXISTS zarr(k TEXT PRIMARY KEY, v BLOB)'
```

Comment:
Should we allow the user to choose a different table name if they want to? That would allow multiple zarr stores inside the same sqlite database.

Reply:
this is what @nbren12 and I have done in #372

Reply:
It's funny you mention this as I had been debating this as well. Actually this shows up in the commit history of this PR too as the store did briefly allow for this.

However ultimately I decided to back that change out because this makes the database vulnerable to SQL injection attacks. Namely we would have to use string formatting to add the table name in and if that name included code it could corrupt the database. This seems problematic especially if the database has other content beside the Zarr Array. We could attempt to sanitize the name provided, but this opens another can of worms.

Ultimately I decided the juice wasn't worth the squeeze. If it's useful, figured people would request it. Besides its one line of the SQL to rename the table.

Reply:
Fair point, forgot about SQL injection. Happy to leave as-is.